### PR TITLE
Item count setting

### PR DESCRIPTION
**Pull request by candry7731**
Add New User setting to adjust the number of item that are loaded inside each folder. 
The default is 60 items

**Changes**
Added Custom voice enabled Number keyboard
Updated the load task to load more if user scrolls down the grid and there is more items to load

### CODE REVIEW

File: settings/settings.json [hunk 1]
```diff
@@ -40,6 +40,13 @@
                         "settingName": "itemgrid.reset",
                         "type": "bool",
                         "default": "true"
+                    },
+                    {
+                        "title": "Items to Load",
+                        "description": "Default Items that are laoded is 60. If you have network issues try reducing this number to help load items faster",
```
Comment (1hitsong) [inline settings/settings.json]: "laoded" misspelling

File: components/ItemGrid/LoadItemsTask2.brs [hunk 2]
```diff
@@ -1,5 +1,11 @@
 sub init()
     m.top.functionName = "loadItems"
+    if get_user_setting("itemgrid.Limit") = "0"
+        m.top.limit = 60
+    else
+        m.top.limit = get_user_setting("itemgrid.Limit")
+    end if
+    print "Load Grid Items Limit: " m.top.limit
```
Comment (1hitsong) [inline components/ItemGrid/LoadItemsTask2.brs]: Debug code

**Review by 1hitsong — CHANGES_REQUESTED** [review]
1st pass, highlevel review

File: components/ItemGrid/LoadItemsTask2.brs (hunk 2 above)
Comment (candry7731) [reply components/ItemGrid/LoadItemsTask2.brs]: removed

**Comment by neilsb** [comment]
When I first load this and go into a library, I get a Type Mismatch error on `reg.write(key, value)` and the client exits.  It's trying to write an integer (60) to the registry, but needs a string.  Suspect maybe just need to make the default value in the settings json.

When I go into the new setting, the keypad works well, but a couple of things
 - When it loads up, the text box is always empty.  Should probably show what the currently set value is - otherwise the user has no way of knowing?
 - It's not clear how to save the new setting.  I assume it's saved automatically when you back out but that leads me to the next point
 - When you press back, the whole settings menu closes.  I would expect it to return you to the settings menu and the user can back out from there

I'm not getting more items loaded in any of my libraries.   Whatever I set the new setting to (e.g. 80) is all the items that are loaded in that library, no matter how much scrolling up and down I do.  The title shows "Movies (80 of 386)" when I highlight he last one the screen, but no further items are loading.

**Comment by candry7731** [comment]
> When I first load this and go into a library, I get a Type Mismatch error on `reg.write(key, value)` and the client exits. It's trying to write an integer (60) to the registry, but needs a string. Suspect maybe just need to make the default value in the settings json.
>  **_I already have a setting saved so I do not see this issue._**
> When I go into the new setting, the keypad works well, but a couple of things
> 
> * When it loads up, the text box is always empty.  Should probably show what the currently set value is - otherwise the user has no way of knowing? **_I have updated this to pull the users save setting, if there is no saved setting it reverts to the default of 60_**
> * It's not clear how to save the new setting.  I assume it's saved automatically when you back out but that leads me to the next point _**when the user hits the back button the value is saved, This is how all other settings are saved including item grid option menus, view, sort and filter**_
> * When you press back, the whole settings menu closes.  I would expect it to return you to the settings menu and the user can back out from there _**This has been fixed**_
> 
> I'm not getting more items loaded in any of my libraries. Whatever I set the new setting to (e.g. 80) is all the items that are loaded in that library, no matter how much scrolling up and down I do. The title shows "Movies (80 of 386)" when I highlight he last one the screen, but no further items are loading. **_This has been fixed. If there are more items to load when you focus on an item it will load the next set of items automatically. No need to do math on what row is focused and how many row in the grid.., This should also speed up the ux when loading more items. _**

File: settings/settings.json [hunk 3]
```diff
@@ -40,9 +40,16 @@
                         "settingName": "itemgrid.reset",
                         "type": "bool",
                         "default": "true"
+                    },
+                    {
+                        "title": "Items to Load",
+                        "description": "Default Items that are loaded is 60. If you have network issues try reducing this number to help load items faster",
+                        "settingName": "itemgrid.Limit",
+                        "type": "integer",
+                        "default": 60
```
Comment (neilsb) [inline settings/settings.json]:
```suggestion
                        "default": "60"
```

Values stored in the registry need to be string values

**Comment by neilsb** [comment]
> I already have a setting saved so I do not see this issue.

Just for future reference, you can uninstall the Dev project from the Roku home screen, which also removes all registry settings.  So when you next deploy the dev package, you can test from a clean install.

> I have updated this to pull the users save setting, if there is no saved setting it reverts to the default of 60

I know this is going to sound picky.  When going into this setting, the Clear and Backspace buttons don't do anything to the original value, and only work if you change the value of the setting in the editor.   I'm just keen we try to have a UI that makes sense to the user, and would work how they would expect from using other roku apps.

Use of the clear icon should (in all cases) clear all data entered into the text box, and by default the cursor should go to the end of the text box so the backspace button removes the last character.

If the user backs out of the setting leaving the entry blank, we could either revert to their previous setting, or set it to the default value.

> when the user hits the back button the value is saved, This is how all other settings are saved including item grid option menus, view, sort and filter

OK - this can be left as is currently.   I think I may look at re-working this later on so that when they go into the setting, it shows the value selected (i.e. 60) and has a single button to "Edit".  This then brings up your pin pad letting them change it, and have a "OK" and "Cancel" button in that popup.   This is how I imagine the string values working too when I get round to implementing that.

> If there are more items to load when you focus on an item it will load the next set of items automatically. No need to do math on what row is focused and how many row in the grid.., This should also speed up the ux when loading more items. 

I'm not sure about this to be honest.  With the default values, by the time you move to the right hand item on the top row, you could have loaded 360 items, even though only 12 will be shown.

There are essentially 2 options when displaying grids like this.   

 - Load on Demand
 - Load Everything

The existing way is Load on Demand.  As the user approaches the end of what is already loaded, load some more, with the aim of reducing network traffic / server load.

Load everything is another valid choice, and can be done in a smart way by loading in chunks rather than loading everything in one go.

I'm not sure the proposed way is either of those; arbitrarily loading each time the user changes selection.  If a load everything approach was desired, it would make sense the trigger the load of the next set of data immediately when the load of the current set is done.  This would be the _speed up the ux_ the most, not taking into account any delay Roku might have managing data for very large libraries.

With load on demand value set at a reasonable value, and the filter, search and alpha picker, I'm not sure there's a need to preload large amounts of data that likely not be used.

Would be interested to get other people's thoughts though.

**Comment by candry7731** [comment]
@neilsb 
I updated the cursor position so the clear and backspace buttons work as intended

The thought behind this setting is to allow the user to decide how they want the grid to load. If they want to load a 1,000 items at once they can by changing the setting. If they only want to load 10 items at a time they can. 

Using the row focus does not work if the user sets the item count lower that the rows that we are trying to gauge when to load more data. For instance if the user set to load 10 items at a time it will only load 1.5 rows and thus will never load more data because we are waiting for the user to be with x amount of rows until the end...

**Comment by cewert** [comment]
> Would be interested to get other people's thoughts though.

Our priority right now is mainly fixing bugs, so I don't see us looking at this anytime soon. I'm thinking we should put this on the back burner until we can give it the attention it deserves. 

There's also some merge conflicts that need to be resolved. Marking this as draft for now

**Comment by candry7731** [comment]
Merge conflicts are completed.

File: settings/settings.json [hunk 4]
```diff
@@ -106,9 +106,16 @@
                         "settingName": "itemgrid.reset",
                         "type": "bool",
                         "default": "true"
+                    },
+                    {
+                        "title": "Items to Load",
+                        "description": "Default Items that are loaded is 60. If you have network issues try reducing this number to help load items faster",
```
Comment (alanazar) [inline settings/settings.json:112]: Adding a full-stop to the end to keep consistency with the rest of the updated settings file
"description": "Default Items that are loaded is 60. If you have network issues try reducing this number to help load items faster.",

Comment (alanazar) [inline settings/settings.json:112]: Making settings.json consistent with the other user settings
```suggestion
                        "description": "Default Items that are loaded is 60. If you have network issues try reducing this number to help load items faster.",
```

File: locale/en_US/translations.ts [hunk 5]
```diff
@@ -573,6 +573,16 @@
         <source>There was an error authenticating via Quick Connect.</source>
         <translation>There was an error authenticating via Quick Connect.</translation>
     </message>
+    <message>
+        <source>Items to Load</source>
+        <translation>Items to Load</translation>
+        <extracomment>UI -> Media Grid -> Item Title in user setting screen.</extracomment>
+    </message>
+    <message>
+        <source>Default Items that are loaded is 60. If you have network issues try reducing this number to help load items faster</source>
```
Comment (alanazar) [inline locale/en_US/translations.ts:582]:
```suggestion
        <source>Default Items that are loaded is 60. If you have network issues try reducing this number to help load items faster.</source>
```

File: locale/en_US/translations.ts [hunk 6]
```diff
@@ -573,6 +573,16 @@
...
+        <translation>Default Items that are loaded is 60. If you have network issues try reducing this number to help load items faster</translation>
```
Comment (alanazar) [inline locale/en_US/translations.ts:583]:
```suggestion
        <translation>Default Items that are loaded is 60. If you have network issues try reducing this number to help load items faster.</translation>
```

**Pull request closed without merging**